wgpu/src/lib.rs

```diff
@@ -3367,6 +3420,32 @@ impl Queue {
         Context::queue_write_buffer(&*self.context, &self.id, &buffer.id, offset, data)
     }
 
+    /// Schedule a data write into `buffer` starting at `offset` via the returned [QueueWriteBufferView].
+    ///
+    /// The returned value can be dereferenced to a `&mut [u8]`; dereferencing it to a `&[u8]` panics!
+    ///
+    /// This method is intended to have low performance costs.
+    /// As such, the write is not immediately submitted, and instead enqueued
+    /// internally to happen at the start of the next `submit()` call.
+    ///
+    /// This method fails if `size` is greater than the size of `buffer` starting at `offset`.
+    #[must_use]
+    pub fn write_buffer_with<'a>(
+        &'a self,
+        buffer: &'a Buffer,
+        offset: BufferAddress,
+        size: BufferSize,
+    ) -> QueueWriteBufferView<'a> {
+        Context::queue_validate_write_buffer(&*self.context, &self.id, &buffer.id, offset, size);
```

Comment:
Why is this a separate call through the context and not just part of queue_create_staging_buffer?

Reply:
I thought it would be appropriate to call it here in `write_buffer_with` since `queue_validate_write_buffer` and `queue_create_staging_buffer` don't share any parameters/functionality.

Reply:
I'd prefer this to be a single call so we don't need to pay the cost of grabbing locks etc again

Reply:
The only lock we could reuse is the device one (one being read; for `queue_validate_write_buffer`, the other write; for `queue_create_staging_buffer`) but you mentioned previously that we'll be able to avoid the locks for the validation part (where we only need to read the size and usage of the dst buffer).

Reply:
Alright, this is good enough for now, I need to revisit write_buffer at some point in the future anyway, so I'll look at it more in depth then.